tests/cpp/transforms/determine_ad_stack_size_test.cpp

```diff
@@ -0,0 +1,154 @@
+
+namespace taichi {
+namespace lang {
+
+class DetermineAdStackSizeTest : public ::testing::Test {
+ protected:
+  void SetUp() override {
+    prog_ = std::make_unique<Program>();
+    prog_->materialize_runtime();
+  }
+
+  std::unique_ptr<Program> prog_;
+};
+
+TEST_F(DetermineAdStackSizeTest, Basic) {
+  IRBuilder builder;
+  auto *stack =
+      builder.create_ad_stack(get_data_type<int>(), 0 /*adaptive size*/);
+  builder.ad_stack_push(stack, builder.get_int32(1));
+  builder.ad_stack_push(stack, builder.get_int32(2));
+  builder.ad_stack_push(stack, builder.get_int32(3));
+  builder.ad_stack_pop(stack);
+  builder.ad_stack_pop(stack);
+  builder.ad_stack_push(stack, builder.get_int32(4));
+  builder.ad_stack_push(stack, builder.get_int32(5));
+  builder.ad_stack_push(stack, builder.get_int32(6));
+  // stack contains [1, 4, 5, 6] now
+  builder.ad_stack_pop(stack);
+  builder.ad_stack_pop(stack);
+  builder.ad_stack_push(stack, builder.get_int32(7));
+
+  auto *stack2 =
+      builder.create_ad_stack(get_data_type<int>(), 0 /*adaptive size*/);
+  builder.ad_stack_push(stack2, builder.get_int32(8));
+
+  auto ir = builder.extract_ir();
+  ASSERT_TRUE(ir->is<Block>());
+  auto *ir_block = ir->as<Block>();
+  irpass::type_check(ir_block, CompileConfig());
+  EXPECT_EQ(ir_block->size(), 22);
+
+  EXPECT_EQ(stack->max_size, 0);
+  EXPECT_EQ(stack2->max_size, 0);
+  irpass::determine_ad_stack_size(ir_block, CompileConfig());
+  EXPECT_EQ(stack->max_size, 4);
+  EXPECT_EQ(stack2->max_size, 1);
+}
+
+TEST_F(DetermineAdStackSizeTest, Loop) {
+  IRBuilder builder;
+  auto *stack =
+      builder.create_ad_stack(get_data_type<int>(), 0 /*adaptive size*/);
+  auto *loop = builder.create_range_for(/*begin=*/builder.get_int32(0),
+                                        /*end=*/builder.get_int32(10));
+  {
+    auto _ = builder.get_loop_guard(loop);
+    builder.ad_stack_push(stack, builder.get_int32(1));
+    builder.ad_stack_pop(stack);
+  }
+
+  auto ir = builder.extract_ir();
+  ASSERT_TRUE(ir->is<Block>());
+  auto *ir_block = ir->as<Block>();
+  irpass::type_check(ir_block, CompileConfig());
+  EXPECT_EQ(irpass::analysis::count_statements(ir_block), 7);
+
+  EXPECT_EQ(stack->max_size, 0);
+  irpass::determine_ad_stack_size(ir_block, CompileConfig());
+  EXPECT_EQ(stack->max_size, 1);
+}
+
+TEST_F(DetermineAdStackSizeTest, LoopInfeasible) {
+  IRBuilder builder;
+  auto *stack =
+      builder.create_ad_stack(get_data_type<int>(), 0 /*adaptive size*/);
+  auto *loop = builder.create_range_for(/*begin=*/builder.get_int32(0),
+                                        /*end=*/builder.get_int32(100));
+  {
+    auto _ = builder.get_loop_guard(loop);
+    builder.ad_stack_push(stack, builder.get_int32(1));
+  }
+
+  auto ir = builder.extract_ir();
+  ASSERT_TRUE(ir->is<Block>());
+  auto *ir_block = ir->as<Block>();
+  irpass::type_check(ir_block, CompileConfig());
+  EXPECT_EQ(irpass::analysis::count_statements(ir_block), 6);
```

Comment:
Is this important? (It's hard to see why there are 6 stmts)

Reply:
I wrote this only because other tests have them... I'll remove it.